NumPy fails to build on aarch64

The master branch currently fails to build on aarch64/ARMv8 with the following:

```
aarch64-linux-gnu-gcc: numpy/core/src/npymath/ieee754.c
In file included from numpy/core/src/private/npy_config.h:6:0,
                 from numpy/core/src/npymath/npy_math_common.h:8,
                 from numpy/core/src/npymath/ieee754.c.src:7:
numpy/core/include/numpy/npy_cpu.h:106:6: error: #error Unknown CPU, please report this to numpy maintainers with information about your platform (OS, CPU and compiler)
     #error Unknown CPU, please report this to numpy maintainers with \
      ^
In file included from numpy/core/src/private/npy_config.h:6:0,
                 from numpy/core/src/npymath/npy_math_common.h:8,
                 from numpy/core/src/npymath/ieee754.c.src:7:
numpy/core/include/numpy/npy_cpu.h:106:6: error: #error Unknown CPU, please report this to numpy maintainers with information about your platform (OS, CPU and compiler)
     #error Unknown CPU, please report this to numpy maintainers with \
      ^
error: Command "aarch64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -Inumpy/core/src/npymath -Inumpy/core/include -Inumpy/core/include/numpy -Inumpy/core/src/private -Inumpy/core/src -Inumpy/core -Inumpy/core/src/npymath -Inumpy/core/src/multiarray -Inumpy/core/src/umath -Inumpy/core/src/npysort -I/usr/include/python3.5m -Inumpy/core/src/private -Inumpy/core/src/npymath -Inumpy/core/src/private -Inumpy/core/src/npymath -Inumpy/core/src/private -Inumpy/core/src/npymath -c numpy/core/src/npymath/ieee754.c -o build/temp.linux-aarch64-3.5/numpy/core/src/npymath/ieee754.o -MMD -MF build/temp.linux-aarch64-3.5/numpy/core/src/npymath/ieee754.o.d" failed with exit status 1
```

I believe this is a result of the new detection logic introduced in #11016.  It seems as if `__arm__` and `__ARMEL__` may not be defined on some? aarch64 platforms.  Rather it may be necessary to use `__aarch64__` and `__AARCH64EL__`.

--- a/numpy/core/include/numpy/npy_cpu.h
+++ b/numpy/core/include/numpy/npy_cpu.h
@@ -63,8 +63,8 @@
     #define NPY_CPU_HPPA
 #elif defined(__alpha__)
     #define NPY_CPU_ALPHA
-#elif defined(__arm__)
-    #if defined(__ARMEB__)
+#elif defined(__arm__) || defined(__aarch64__)
+    #if defined(__ARMEB__) || defined(__AARCH64EB__)
         #if defined(__ARM_32BIT_STATE)
             #define NPY_CPU_ARMEB_AARCH32
         #elif defined(__ARM_64BIT_STATE)
@@ -72,7 +72,7 @@
         #else
             #define NPY_CPU_ARMEB
         #endif
-    #elif defined(__ARMEL__)
+    #elif defined(__ARMEL__) || defined(__AARCH64EL__)
         #if defined(__ARM_32BIT_STATE)
             #define NPY_CPU_ARMEL_AARCH32
         #elif defined(__ARM_64BIT_STATE)

--- a/numpy/core/src/multiarray/dragon4.c
+++ b/numpy/core/src/multiarray/dragon4.c
@@ -2736,7 +2736,6 @@ Dragon4_PrintFloat_IEEE_binary128(
         return 0;
     }
 
-    /* Assumes little-endian !!! */
     mantissa_hi = val128.hi & bitmask_u64(48);
     mantissa_lo = val128.lo;
     floatExponent = (val128.hi >> 48) & bitmask_u32(15);
@@ -2824,7 +2823,7 @@ Dragon4_PrintFloat_IEEE_binary128_le(
 #if defined(HAVE_LDOUBLE_IEEE_QUAD_BE)
 /*
  * This function is untested, very few, if any, architectures implement
- * big endien IEEE binary128 floating point.
+ * big endian IEEE binary128 floating point.
  */
 static npy_uint32
 Dragon4_PrintFloat_IEEE_binary128_be(

--- a/numpy/core/src/multiarray/dragon4.c
+++ b/numpy/core/src/multiarray/dragon4.c
@@ -2707,11 +2707,9 @@ Dragon4_PrintFloat_Intel_extended128(
  * mantissa: 112 bits
  *
  * Currently binary128 format exists on only a few CPUs, such as on the POWER9
- * arch. Because of this, this code has not been tested. I am not sure if the
- * arch also supports uint128, and C does not seem to support int128 literals.
- * So we use uint64 to do manipulation. Unfortunately this means we are endian
- * dependent. Assume little-endian for now, can fix later once binary128
- * becomes more common.
+ * arch or aarch64. Because of this, this code has not been extensively tested.
+ * I am not sure if the arch also supports uint128, and C does not seem to
+ * support int128 literals. So we use uint64 to do manipulation.
  */
 static npy_uint32
 Dragon4_PrintFloat_IEEE_binary128(
@@ -2824,6 +2822,10 @@ Dragon4_PrintFloat_IEEE_binary128_le(
 #endif /* HAVE_LDOUBLE_IEEE_QUAD_LE */
 
 #if defined(HAVE_LDOUBLE_IEEE_QUAD_BE)
+/*
+ * This function is untested, very few, if any, architectures implement
+ * big endien IEEE binary128 floating point.
+ */
 static npy_uint32
 Dragon4_PrintFloat_IEEE_binary128_be(
     Dragon4_Scratch *scratch, npy_float128 *value, Dragon4_Options *opt)

--- a/numpy/core/src/multiarray/dragon4.c
+++ b/numpy/core/src/multiarray/dragon4.c
@@ -2714,7 +2714,7 @@ Dragon4_PrintFloat_Intel_extended128(
  * becomes more common.
  */
 static npy_uint32
-Dragon4_PrintFloat_IEEE_binary128(
+Dragon4_PrintFloat_IEEE_binary128_le(
     Dragon4_Scratch *scratch, npy_float128 *value, Dragon4_Options *opt)
 {
     FloatUnion128 buf128;

--- a/numpy/core/src/multiarray/dragon4.c
+++ b/numpy/core/src/multiarray/dragon4.c
@@ -2743,10 +2743,10 @@ Dragon4_PrintFloat_IEEE_binary128_le(
     }
 
     /* Assumes little-endian !!! */
-    mantissa_hi = buf128.integer.a & bitmask_u64(48);
-    mantissa_lo = buf128.integer.b;
-    floatExponent = (buf128.integer.a >> 48) & bitmask_u32(15);
-    floatSign = buf128.integer.a >> 63;
+    mantissa_hi = buf128.integer.b & bitmask_u64(48);
+    mantissa_lo = buf128.integer.a;
+    floatExponent = (buf128.integer.b >> 48) & bitmask_u32(15);
+    floatSign = buf128.integer.b >> 63;
 
     /* output the sign */
     if (floatSign != 0) {

--- a/numpy/core/src/multiarray/dragon4.c
+++ b/numpy/core/src/multiarray/dragon4.c
@@ -2698,7 +2698,7 @@ Dragon4_PrintFloat_Intel_extended128(
 }
 #endif /* HAVE_LDOUBLE_INTEL_EXTENDED_16_BYTES_LE */
 
-#if defined(HAVE_LDOUBLE_IEEE_QUAD_LE)
+#if defined(HAVE_LDOUBLE_IEEE_QUAD_LE) || defined(HAVE_LDOUBLE_IEEE_QUAD_BE)
 /*
  * IEEE binary128 floating-point format
  *
@@ -2714,11 +2714,9 @@ Dragon4_PrintFloat_Intel_extended128(
  * becomes more common.
  */
 static npy_uint32
-Dragon4_PrintFloat_IEEE_binary128_le(
-    Dragon4_Scratch *scratch, npy_float128 *value, Dragon4_Options *opt)
+Dragon4_PrintFloat_IEEE_binary128(
+    Dragon4_Scratch *scratch, FloatVal128 val128, Dragon4_Options *opt)
 {
-    FloatUnion128 buf128;
-
     char *buffer = scratch->repr;
     npy_uint32 bufferSize = sizeof(scratch->repr);
     BigInt *bigints = scratch->bigints;
@@ -2731,8 +2729,6 @@ Dragon4_PrintFloat_IEEE_binary128_le(
     npy_bool hasUnequalMargins;
     char signbit = '\0';
 
-    buf128.floatingPoint = *value;
-
     if (bufferSize == 0) {
         return 0;
     }
@@ -2743,10 +2739,10 @@ Dragon4_PrintFloat_IEEE_binary128_le(
     }
 
     /* Assumes little-endian !!! */
-    mantissa_hi = buf128.integer.b & bitmask_u64(48);
-    mantissa_lo = buf128.integer.a;
-    floatExponent = (buf128.integer.b >> 48) & bitmask_u32(15);
-    floatSign = buf128.integer.b >> 63;
+    mantissa_hi = val128.hi & bitmask_u64(48);
+    mantissa_lo = val128.lo;
+    floatExponent = (val128.hi >> 48) & bitmask_u32(15);
+    floatSign = val128.hi >> 63;
 
     /* output the sign */
     if (floatSign != 0) {
@@ -2810,7 +2806,40 @@ Dragon4_PrintFloat_IEEE_binary128_le(
     return Format_floatbits(buffer, bufferSize, bigints, exponent,
                             signbit, mantissaBit, hasUnequalMargins, opt);
 }
+
+#if defined(HAVE_LDOUBLE_IEEE_QUAD_LE)
+static npy_uint32
+Dragon4_PrintFloat_IEEE_binary128_le(
+    Dragon4_Scratch *scratch, npy_float128 *value, Dragon4_Options *opt)
+{
+    FloatVal128 val128;
+    FloatUnion128 buf128;
+
+    buf128.floatingPoint = *value;
+    val128.lo = buf128.integer.a;
+    val128.hi = buf128.integer.b;
+
+    return Dragon4_PrintFloat_IEEE_binary128(scratch, val128, opt);
+}
 #endif /* HAVE_LDOUBLE_IEEE_QUAD_LE */
+
+#if defined(HAVE_LDOUBLE_IEEE_QUAD_BE)
+static npy_uint32
+Dragon4_PrintFloat_IEEE_binary128_be(
+    Dragon4_Scratch *scratch, npy_float128 *value, Dragon4_Options *opt)
+{
+    FloatVal128 val128;
+    FloatUnion128 buf128;
+
+    buf128.floatingPoint = *value;
+    val128.lo = buf128.integer.b;
+    val128.hi = buf128.integer.a;
+
+    return Dragon4_PrintFloat_IEEE_binary128(scratch, val128, opt);
+}
+#endif /* HAVE_LDOUBLE_IEEE_QUAD_BE */
+
+#endif /* HAVE_LDOUBLE_IEEE_QUAD_LE | HAVE_LDOUBLE_IEEE_BE*/
 
 #if (defined(HAVE_LDOUBLE_IBM_DOUBLE_DOUBLE_LE) || \
      defined(HAVE_LDOUBLE_IBM_DOUBLE_DOUBLE_BE))

